### PR TITLE
Add Snaps Keyring API documentation

### LINKED ISSUE #846: Document the Snaps Keyring API
AC:

- Reference documentation is updated to include the `snap_manageAccounts` method
- Concepts section has a new "keyring" (name TBD) subsection with details on the API and implementation
- How-To section has a new "Add custom accounts to MetaMask" subsection which mentions the keyring API and has links to the other two sections where necessary

TODO:

- Figure out the most appropriate to leverage Docusaurus versioning have the keyring API documented, but "segregated" from the other docs, since it won't be part of the V1 release.

### PR DESCRIPTION
**Pull request by ziad-saab**
Fixes #846

### CODE REVIEW

**Comment by ziad-saab** [comment]
@SocketSecurity ignore-all

File: snaps/how-to/integrate-custom-evm-accounts.md [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+
+![Keyring snap account creation flow](../assets/keyring/create-account-flow.png)
+
+The MetaMask account selection modal has an option called "Add snap account":
+
+![add snap account option](../assets/keyring/add-snap-account.png)
+
+This option will show a list of keyring snaps, but will ultimately redirect the user to the companion Dapp for your Keyring snap. That is, a Dapp that contains all the UI to configure and manage the Keyring snap.
+
+On that Dapp, you'll present a custom user interface allowing the user configure their custom EVM account. The Dapp will make use of the [`KeyringSnapRpcClient`'s `createAccount` method](../reference/keyring-api/classes/KeyringSnapRpcClient.md#createaccount), which will call your `Keyring` class' method of the same name. An example of this can be found in the [Simple keyring snap companion Dapp](https://github.com/MetaMask/snap-simple-keyring/blob/d3f7f0156c59059c995fea87f90a3d0ad3a4c135/packages/site/src/pages/index.tsx#L136).
```
Comment (plasmacorral) [inline snaps/how-to/integrate-custom-evm-accounts.md:51]: Getting a 404 @ https://github.com/MetaMask/metamask-docs/blob/zs/keyring-doc/snaps/reference/keyring-api/classes/KeyringSnapRpcClient.md#createaccount as the `keyring-api` folder does not seem to exist.

File: snaps/how-to/integrate-custom-evm-accounts.md [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+Your Keyring snap will implement the simple flow if it's able to sign transactions directly. This would be the case if the snap doesn't need a third-party such as a hardware key or a second account's signature, as would be the case for a threshold signature scheme. The flow would look like this:
+
+![Synchronous signing flow](../assets/keyring/synchronous-flow.png)
+
+For a full example of a simple keyring snap, see [`snap-simple-keyring` on GitHub](https://github.com/MetaMask/snap-simple-keyring).
+
+
+The flow starts when a Dapp calls a method such as `personal_sign` or `eth_signTransaction`, or when the user initiates a new funds transfer from the MetaMask UI. At that point, MetaMask will detect that this interaction is requested for an account controlled by your Keyring snap.
+
+After the user approves the transaction in the UI, MetaMask will call the `submitRequest` method of your `Keyring` class. `submitRequest` will receive the original RPC request, and will need to return a [`SubmitRequestResponse`](../reference/keyring-api/modules.md#submitrequestresponsestruct) with `pending` set to `false`, and `result` set to the requested signature.
```
Comment (plasmacorral) [inline snaps/how-to/integrate-custom-evm-accounts.md:68]: Also a 404 @ https://github.com/MetaMask/metamask-docs/blob/zs/keyring-doc/snaps/reference/keyring-api/modules.md#submitrequestresponsestruct

File: snaps/how-to/integrate-custom-evm-accounts.md [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+
+![Asynchronous signing flow](../assets/keyring/asynchronous-flow.png)
+
+The flow starts the same way: a request to sign a transaction or arbitrary data is initiated by a Dapp or by the user. After approval, your snap's `submitRequest` method is called. 
+
+Since your snap won't answer the request directly, it should store the pending request in its internal state using [`snap_manageState`](../reference/rpc-api.md#snap_managestate). This list of pending requests should be returned when the `listRequests` or `getRequest` methods of your `Keyring` class are called.
+
+After storing the pending request, your snap should direct the user to a "companion Dapp" – a Dapp that serves as UI for the snap – where the rest of the flow can continue. Your snap can do this by creating a pop-up using [`snap_dialog`](../reference/rpc-api.md#snap_dialog) instructing the user to go to the companion Dapp's URL.
+
+The Dapp will list your snap's pending requests using an RPC call facilitated by the [`KeyringSnapRpcClient`'s `listRequests` method](../reference/keyring-api/classes/KeyringSnapRpcClient.md#listrequests). The user can then act on those requests using whatever process applies to the snap.
```
Comment (plasmacorral) [inline snaps/how-to/integrate-custom-evm-accounts.md:86]: 404 @ https://github.com/MetaMask/metamask-docs/blob/zs/keyring-doc/snaps/reference/keyring-api/classes/KeyringSnapRpcClient.md#listrequests

File: snaps/how-to/integrate-custom-evm-accounts.md [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+
+Once the signing process is completed, the companion Dapp will resolve the request using the [`KeyringSnapRpcClient`'s `approveRequest` method](../reference/keyring-api/classes/KeyringSnapRpcClient.md#approverequest), which will call the snap's `Keyring` method of the same name. This method receives the request's ID, as well as the final result.
```
Comment (plasmacorral) [inline snaps/how-to/integrate-custom-evm-accounts.md:88]: 404 https://github.com/MetaMask/metamask-docs/blob/zs/keyring-doc/snaps/reference/keyring-api/classes/KeyringSnapRpcClient.md#approverequest

File: snaps/how-to/integrate-custom-evm-accounts.md [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+    "snap_manageAccounts": {}
+  }
+}
+```
+
+## Exposing the `Keyring` class as JSON-RPC endpoints
+
+Once you've written all the business logic for your custom EVM account type, the next step is to expose it as a JSON-RPC API. Exposing a JSON-RPC API in Snaps is done by [exporting an `onRpcRequest` function](../reference/exports.md#onrpcrequest) from the snap package.
+
+The `@metamask/keyring-api` package provides a helper called [`handleKeyringRequest`](../reference/keyring-api/modules.md#handlekeyringrequest). This helper takes an instance of your `Keyring` class and a request object. It responds to requests where the `method` is of type `keyring_*`, and throws a [`MethodNotSupportedError`](../reference/keyring-api/classes/MethodNotSupportedError.md) if it doesn't recognize the request method.
```
Comment (plasmacorral) [inline snaps/how-to/integrate-custom-evm-accounts.md:110]: 404 https://github.com/MetaMask/metamask-docs/blob/zs/keyring-doc/snaps/reference/keyring-api/classes/MethodNotSupportedError.md

File: snaps/how-to/integrate-custom-evm-accounts.md [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+  keyringHandler,
+  customHandler,
+);
+```
+
+## Using the keyring API from a Dapp
+
+As you're building a companion Dapp to provide a user interface for your Keyring snap, you'll need to interact with your snap's JSON-RPC API. While you could do this by making regular RPC calls using [`wallet_invokeSnap`](../reference/rpc-api.md#wallet_invokesnap), we provide a more elegant solution.
+
+From your Dapp, you can use the [`KeyringSnapRpcClient`](../reference/keyring-api/classes/KeyringSnapRpcClient.md) from the `@metamask/keyring-api` package, like so:
```
Comment (plasmacorral) [inline snaps/how-to/integrate-custom-evm-accounts.md:184]: 404: https://github.com/MetaMask/metamask-docs/blob/zs/keyring-doc/snaps/reference/keyring-api/classes/KeyringSnapRpcClient.md

File: snaps/how-to/integrate-custom-evm-accounts.md (hunk 1 above)
Comment (ziad-saab) [reply snaps/how-to/integrate-custom-evm-accounts.md:51]: That's normal, as it gets created at build time from a Git submodule!

File: snaps/how-to/integrate-custom-evm-accounts.md (hunk 2 above)
Comment (ziad-saab) [reply snaps/how-to/integrate-custom-evm-accounts.md:68]: That's normal, as it gets created at build time from a Git submodule!

File: snaps/concepts/keyring-api.md [hunk 7]
```diff
@@ -0,0 +1,9 @@
+---
+description: Learn about the Snaps Keyring API.
+---
+
+# Keyring API
+
+The Keyring API brings deeper integration for custom EVM accounts inside MetaMask. Before the Keyring API, custom EVM accounts such as ERC-4337, MPC accounts, etc. needed to be shown separately, in a companion Dapp. Now these custom accounts can live alongside regular MetaMask accounts in the UI. Dapps can connect to them using `eth_requestAccounts`, and seamlessly interact with them using `eth_sendTransaction`, `personal_sign`, etc.
```
Comment (danroc) [inline snaps/concepts/keyring-api.md:7]: Sadly, 4337 accounts are not supported (yet)

File: snaps/assets/keyring/accounts-ui.png
Comment (Montoya) [inline snaps/assets/keyring/accounts-ui.png:1]: This still refers to ERC-4337?

**Comment by Montoya** [comment]
I would prefer that Keyrings API is last on this list
Screenshot 2023-09-15 at 2 23 04 PM

**Comment by Montoya** [comment]
This formatting seems wrong? Is there a better way to format these structs? 
Screenshot 2023-09-15 at 2 21 02 PM

**Review by Montoya — CHANGES_REQUESTED** [review]

File: snaps/assets/keyring/accounts-ui.png
Comment (Montoya) [reply snaps/assets/keyring/accounts-ui.png:1]: Aside from this request, see my other two comments above.

**Comment by ziad-saab** [comment]
> This formatting seems wrong? Is there a better way to format these structs? Screenshot 2023-09-15 at 2 21 02 PM

There's a beta version of the typedoc plugin that could work, I will try it and report back.

File: snaps/assets/keyring/accounts-ui.png
Comment (ziad-saab) [reply snaps/assets/keyring/accounts-ui.png:1]: fixed

**Comment by ziad-saab** [comment]
@Montoya the new version of the Typedoc plugin looks a lot better:

https://docs.metamask.io/zs/keyring-doc/snaps/reference/keyring-api/Variables/variable.KeyringAccountStruct/

Unfortunately I noticed a bug where some of the types don't get displayed:

https://docs.metamask.io/zs/keyring-doc/snaps/reference/keyring-api/Type%20Aliases/type-alias.Keyring/

I already submitted an issue on the typedoc plugin's GitHub: https://github.com/tgreyuk/typedoc-plugin-markdown/issues/470

What do you think? Should we keep the older, ugly version? Or the newer prettier version with the knowledge that it will be fixed soon?

**Comment by ziad-saab** [comment]
> I would prefer that Keyring API is last on this list Screenshot 2023-09-15 at 2 23 04 PM

fixed

**Comment by Montoya** [comment]
> What do you think? Should we keep the older, ugly version? Or the newer prettier version with the knowledge that it will be fixed soon?

The new version still seems better

**Comment by Montoya** [comment]
It looks like some of the examples are rendering badly because something is broken or cut off. I see some broken tags mixed in. 

Screenshot 2023-09-15 at 7 37 29 PM

**Comment by ziad-saab** [comment]
> It looks like some of the examples are rendering badly because something is broken or cut off. I see some broken tags mixed in.
> Screenshot 2023-09-15 at 7 37 29 PM

Submitted a new issue here >> https://github.com/tgreyuk/typedoc-plugin-markdown/issues/473

As agreed during standup, we'll merge as-is since there are no other issues.

**Review by Montoya — APPROVED** [review]
LGTM

**Comment by alexandratran** [comment]
Will review this this week.

**Review by alexandratran — APPROVED** [review]
Made some edits and approving for now, but we need to edit the `keyring-api` submodule and update the flow/architecture diagrams in the future.

**Pull request merged**